File: LeetCode/1557. Minimum Number of Vertices to Reach All Nodes/solution.py
"""
66 / 66 test cases passed.
Runtime: 140 ms
Memory Usage: 36 MB
"""
class Solution:
    def findSmallestSetOfVertices(self, n: int, edges: List[List[int]]) -> List[int]:
        indegree = [0] * n
        for edge in edges:
            indegree[edge[1]] = True
        ans = []
        for i, d in enumerate(indegree):
            if d == 0:
                ans.append(i)
        return ans

File: LeetCode/1557. Minimum Number of Vertices to Reach All Nodes/solution.cpp
/**
 * 66 / 66 test cases passed.
 * Runtime: 212 ms
 * Memory Usage: 91.4 MB 
 */
class Solution {
public:
    vector<int> findSmallestSetOfVertices(int n, vector<vector<int>>& edges) {
        vector<bool> hasIndegree(n);
        for (auto &edge : edges) {
            hasIndegree[edge[1]] = true;
        }
        vector<int> ans;
        for (int i = 0; i < n; ++ i) {
            if (!hasIndegree[i]) {
                ans.emplace_back(i);
            }
        }
        return ans;
    }
};
